compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,304 @@
+package dotty.tools
+package dotc
+package semanticdb
+
+import core.Symbols._
+import core.Contexts.Context
+import core.Types._
+import core.Annotations.Annotation
+import core.Flags
+import core.Names.Name
+import core.StdNames.tpnme
+import ast.tpd._
+
+import collection.mutable
+
+import dotty.tools.dotc.{semanticdb => s}
+
+class TypeOps:
+  import SymbolScopeOps._
+  private val paramRefSymtab = mutable.Map[(LambdaType, Name), Symbol]()
+  private val refinementSymtab = mutable.Map[(RefinedType, Name), Symbol]()
+  given typeOps: TypeOps = this
+  extension (tpe: Type)
+    def toSemanticSig(using LinkMode, Context, SemanticSymbolBuilder)(sym: Symbol): s.Signature =
+      def enterParamRef(tpe: Type): Unit =
+        tpe match {
+          case lam: LambdaType =>
+            paramRefSymtab((lam, sym.name)) = sym
+
+          // for class constructor
+          // class C[T] { ... }
+          case cls: ClassInfo if sym.info.isInstanceOf[LambdaType] =>
+            val lam = sym.info.asInstanceOf[LambdaType]
+            cls.cls.typeParams.foreach { param =>
+              paramRefSymtab((lam, param.name)) = param
+            }
+
+          // type X[T] = ...
+          case tb: TypeBounds =>
+            enterParamRef(tb.lo)
+            enterParamRef(tb.hi)
+
+          case _ => ()
+        }
+
+      def enterRefined(tpe: Type): Unit =
+        tpe match {
+          case refined: RefinedType =>
```

Comment:
This is needed because, for example the `tpe` is instance of `RecType`, it recursively call `enterRefined(rec.parent)` and `RecType#parent` should be always `RefinedType`. (this case is the base case of this recursive function).

Maybe we can refactor it from
```scala
case refined: RefinedType =>
  val key = (refined, sym.name)
  refinementSymtab(key) = sym
```

to 

```scala
case refined: RefinedType =>
  enterRefined(refined)
```

Reply:
https://github.com/lampepfl/dotty/pull/12885/commits/3cad5f33d80a7bb7a6c8001bf4531ccb61c12d9d